File: leetcode654/leetcode654.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 * };
 */
class Solution {
public:
    TreeNode* constructMaximumBinaryTree(vector<int>& nums) {
        if (nums.empty()) return nullptr;
        auto max_iter = max_element(nums.begin(), nums.end());
        int max_val = *max_iter;
        int max_idx = distance(nums.begin(), max_iter);
        TreeNode* root = new TreeNode(max_val);
        vector<int> left(nums.begin(), nums.begin() + max_idx);
        vector<int> right(nums.begin() + max_idx + 1, nums.end());
        root->left = constructMaximumBinaryTree(left);
        root->right = constructMaximumBinaryTree(right);
        return root;
    }
};

File: leetcode654/leetcode654.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode constructMaximumBinaryTree(int[] nums) {
        return build(nums, 0, nums.length);
    }

    private TreeNode build(int[] nums, int start, int end) {
        if (start >= end) return null;
        int maxIdx = start;
        for (int i = start; i < end; i++) {
            if (nums[i] > nums[maxIdx]) {
                maxIdx = i;
            }
        }
        TreeNode root = new TreeNode(nums[maxIdx]);
        root.left = build(nums, start, maxIdx);
        root.right = build(nums, maxIdx + 1, end);
        return root;
    }
}
